walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,175 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
+  /**
+   * Handles "find a meeting" feature by using the exsisting API to return the times
+   * when the meeting could happen that day, given the meeting information.
+   *
+   * @param events all events happening on a given day including event name, 
+   *                a TimeRange, and collection of attendees.
+   *
+   * @param request details of meeting being requested including requested meeting name, 
+   *                duration, and collection of attendees.
+   *
+   * @return an ArrayList containing all the TimeRanges the request meeting can be scheduled
+   *
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<String> attendees = new ArrayList<String>(request.getAttendees());
+    List<TimeRange> busyTimeRanges = findBusyTimes(events, attendees);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      List<String> optional = new ArrayList<String>(request.getOptionalAttendees());
+      attendees.addAll(optional);
+
+      List<TimeRange> optionalBusyTimeRanges = findBusyTimes(events, attendees);
+      List<TimeRange> optionalMeetingTimes = findMeetingTimes(optionalBusyTimeRanges, duration);
+
+      if(!optionalMeetingTimes.isEmpty()) {
+        return optionalMeetingTimes;
+      } else {
+        return Arrays.asList();
+      }
+
+    }
+
+    return findMeetingTimes(busyTimeRanges, duration);
+  }
+
+  //finds meeting times that don't fall during busy times of the day
+  private List<TimeRange> findMeetingTimes(List<TimeRange> busyTimeRanges, long duration) {
+    if (busyTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(busyTimeRanges, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> busyTimeMerged = mergeBusyTimes(busyTimeRanges);
+    List<TimeRange> sortByEndTime = sortByEnd(busyTimeMerged);
```

Comment:
Why is it possible that busyTimeMerged isn't sorted here? I think your mergeBusyTimes guarantees that the end time of previous time range < the start time of next time range.

Reply:
Oh that's surprising. I thought if there is an all-day event in busyTimeRanges, busyTimeMerged should just contain one time range, which is all day. Am I understanding the things wrong here?

Reply:
No, I think you're right, but the problem I'm having is that when I delete the sort, when I run the tests with optional attendees, it seems to be taking into account the optional attendee's schedule when what the test wants is for the optional attendee to be ignored.

Reply:
In your query() method, when the optional attendees are not empty, you just add all optional attendees to the list of regular attendees. That's why I asked you "what is the difference between optional attendees and regular attendees" question because I don't see how your findMeetingTimes() and findBusyTimes() methods can distinguish between regular and optional attendees.

Perhaps what you intended to do is creating a new list that contains only optional attendees?

Reply:
I see it, I don't think I ever built out the optional attendees function, I just assumed every optional attendee could be ignored in planning a time, but the problem is that some of the tests they get us to build out later some want the optional employee to be taken into account but others don't, I could be interpreting them incorrectly. Here's the screenshot: https://screenshot.googleplex.com/UMB6TYw5hLy.png

Reply:
I realized what the issue that required this sort was, in the two tests that fail without this sort line, I don't enter the events in order of start time which is why I need to sort the times before finding the free times. In all the other cases when I add them in order of start time this issue doesn't occur.

Reply:
I don't think mergeBusyTimes sorts. It is expecting a sorted input.

Reply:
Since you said "I need to sort the times before finding the free times" I am assuming you are talking about sorting "busyTimeRanges" by start time at line 74? That line is necessary because busyTimeRanges isn't guaranteed to be sorted, and if it is not, your merging logic won't work.

My original comment is about the sorting at line 77, "List<TimeRange> sortByEndTime = sortByEnd(busyTimeMerged);" I thought it is weird to have another sort here because:
1. by your definition of mergeBusyTimes, busyTimeMerged should be in sorted order if busyTimeRanges is sorted, so you don't need to sort again.
2. Having two versions of the same list (BusyTimeMerged and SortByEndTime) and passing both of them into findFreeTimes looks weird, because you usually don't need two copies of the same list. It suggests that you are doing something wrong/unnecessary.

Reply:
ok got it! I use the sortByEndTime to check whether there is free time between the end of the last meeting at the end of the day (this check is at the end of the findFreeTime function). I needed this because there could be a meeting that starts the latest but ends earlier than another meeting that starts earlier but ends the latest and if I use the busyTimeMerged list, I would have to have a for loop to go through every single timeRange to find the one that ends the latest since it is sorted by start time, whereas if I used the sort function given, I would be able to just get the last timeRange on that list.